syntax = "proto3";

package posts_service;

message Post {
  int32 id = 1;
  string title = 2;
  string description = 3;
  string created_at = 4;
  string updated_at = 5;
  bool is_private = 6;
  int32 creator_id = 7;
  repeated string tags = 8;
}

message CreatePostRequest {
  string title = 1;
  string description = 2;
  bool is_private = 3;
  int32 creator_id = 4;
  repeated string tags = 5;
}

message CreatePostResponse {
  Post post = 1;
}

message DeletePostRequest {
  int32 id = 1;
  int32 user_id = 2;
}

message DeletePostResponse {

}

message UpdatePostRequest {
  int32 id = 1;
  string title = 2;
  string description = 3;
  bool is_private = 6;
  int32 user_id = 7;
  repeated string tags = 8;
}

message UpdatePostResponse {
  Post post = 1;
}

message GetPostRequest {
  int32 id = 1;
  int32 user_id = 2;
}

message GetPostResponse {
  Post post = 1;
}

message ListPostsRequest {
  int32 page = 1;
  int32 page_size = 2;
  int32 user_id = 3;
}

message ListPostsResponse {
  repeated Post posts = 1;
  int32 total = 2;
  int32 page = 3;
  int32 page_size = 4;
}

service PostsService {
  rpc CreatePost (CreatePostRequest) returns (CreatePostResponse);
  rpc DeletePost (DeletePostRequest) returns (DeletePostResponse);
  rpc UpdatePost (UpdatePostRequest) returns (UpdatePostResponse);
  rpc GetPost (GetPostRequest) returns (GetPostResponse);
  rpc ListPosts (ListPostsRequest) returns (ListPostsResponse);
}